sstables/sstables.cc

```diff
@@ -1183,6 +1183,78 @@ void sstable::set_position_range() {
     _position_range = position_range(pip(min_elements, bound_kind::incl_start), pip(max_elements, bound_kind::incl_end));
 }
 
+future<std::optional<position_in_partition>>
+sstable::find_first_position_in_partition(reader_permit permit, const dht::decorated_key& key, bool reversed) {
+    using position_in_partition_opt = std::optional<position_in_partition>;
+    class position_finder {
+        position_in_partition_opt& _pos;
+        bool _reversed;
+    private:
+        void on_position_found(position_in_partition&& pos) {
+            _pos = std::move(pos);
+        }
+    public:
+        position_finder(position_in_partition_opt& pos, bool reversed) noexcept : _pos(pos), _reversed(reversed) {}
+
+        void consume_new_partition(const dht::decorated_key& dk) {}
+
+        stop_iteration consume(tombstone t) {
+            // Handle case partition contains only a partition_tombstone, so position_in_partition
+            // for this key should be before all rows.
+            on_position_found(position_in_partition::before_all_clustered_rows());
+            return stop_iteration::no;
+        }
+
+        stop_iteration consume(range_tombstone_change&& rt) {
+            on_position_found(std::move(rt).position());
+            return stop_iteration::yes;
+        }
+
+        stop_iteration consume(clustering_row&& cr) {
+            on_position_found(position_in_partition::for_key(std::move(cr.key())));
+            return stop_iteration::yes;
+        }
+
+        stop_iteration consume(static_row&& sr) {
+            on_position_found(position_in_partition(sr.position()));
+            // If reversed == true, we shouldn't stop at static row as we want to find the last row.
+            // We don't want to ignore its position, to handle the case where partition contains only static row.
+            return stop_iteration(!_reversed);
+        }
+
+        stop_iteration consume_end_of_partition() {
+            return stop_iteration::yes;
+        }
+
+        mutation_opt consume_end_of_stream() {
+            return std::nullopt;
+        }
+    };
+
+    auto pr = dht::partition_range::make_singular(key);
+    auto s = get_schema();
+    auto full_slice = s->full_slice();
+    if (reversed) {
+        s = s->make_reversed();
+        full_slice.options.set(query::partition_slice::option::reversed);
+    }
+    auto r = make_reader(s, std::move(permit), pr, full_slice);
```

Comment:
Sadly, `make_reader()` defaults to `mutation_reader::forwarding::yes`. I think it is worth turning off, to avoid over-read.

Reply:
is it relevant for single partition read? mx mutation reader is able to detect when reading from a single partition. 

```
    // We avoid unnecessary lookup for single partition reads thanks to this flag
    bool _single_partition_read = false;
```

but let me turn it off as the optimization is not explicit in the interface. if there were no such a flag, the reader could indeed read ahead past the partition end:

```
            sstable::disk_read_range drr{begin, *end};
            auto last_end = _fwd_mr ? _sst->data_size() : drr.end;
```

Reply:
That param should really default to `::no` but when I changed it so a bunch of tests started failing. And I never got back to finish debugging them.

Reply:
Probably some users of the interface rely on this assumption of fast forwarding enabled by default. 

We can probably:
1) make all users of sstable::make_reader() specify mutation_reader::forwarding::yes explicitly.
2) change the default to mutation_reader::forwarding::no
3) finally, switch to mutation_reader::forwarding::no in users of sstable::make_reader() which will benefit from it

Reply:
Yes, this is what I tried. I don't remember the specifics, but there were like two tests that kept failing and I couldn't find out why. We changed a lot around this path, maybe it wouldn't be as hard now.